app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/NetworkTimelineViewModel.kt

```diff
@@ -0,0 +1,280 @@
+        status.copy(
+            status = status.status.copy(poll = newPoll)
+        ).update()
+    }
+
+    override fun changeExpanded(expanded: Boolean, status: StatusViewData.Concrete) {
+        status.copy(
+            isExpanded = expanded
+        ).update()
+    }
+
+    override fun changeContentHidden(isShowing: Boolean, status: StatusViewData.Concrete) {
+        status.copy(
+            isShowingContent = isShowing
+        ).update()
+    }
+
+    override fun changeContentCollapsed(isCollapsed: Boolean, status: StatusViewData.Concrete) {
+        status.copy(
+            isCollapsed = isCollapsed
+        ).update()
+    }
+
+    override fun removeAllByAccountId(accountId: String) {
+        statusData.removeAll { vd ->
+            val status = vd.asStatusOrNull()?.status ?: return@removeAll false
+            status.account.id == accountId || status.actionableStatus.account.id == accountId
+        }
+        currentSource?.invalidate()
+    }
+
+    override fun removeAllByInstance(instance: String) {
+        statusData.removeAll { vd ->
+            val status = vd.asStatusOrNull()?.status ?: return@removeAll false
+            LinkHelper.getDomain(status.account.url) == instance
+        }
+        currentSource?.invalidate()
+    }
+
+    override fun removeStatusWithId(id: String) {
+        statusData.removeAll { vd ->
+            val status = vd.asStatusOrNull()?.status ?: return@removeAll false
+            status.id == id || status.reblog?.id == id
+        }
+        currentSource?.invalidate()
+    }
+
+    override fun loadMore(placeholderId: String) {
+        viewModelScope.launch {
+            val statusResponse = fetchStatusesForKind(
+                fromId = placeholderId.inc(),
+                uptoId = null,
+                limit = 20
+            )
+
+            val statuses = statusResponse.body()
+            if (!statusResponse.isSuccessful || statuses == null) {
+                Log.w("NetworkTimelineVM", "failed loading statuses", HttpException(statusResponse))
+
+                val index = statusData.indexOfFirst { it is StatusViewData.Placeholder && it.id == placeholderId }
+                statusData[index] = StatusViewData.Placeholder(placeholderId, isLoading = false)
+
+                currentSource?.invalidate()
+                return@launch
+            }
+
+            val activeAccount = accountManager.activeAccount!!
+
+            val data = statuses.map { status ->
+                status.toViewData(
+                    alwaysShowSensitiveMedia = !activeAccount.alwaysShowSensitiveMedia && status.actionableStatus.sensitive,
+                    alwaysOpenSpoiler = activeAccount.alwaysOpenSpoiler
+                )
+            }
+
+            val index = statusData.indexOfFirst { it is StatusViewData.Placeholder && it.id == placeholderId }
+            statusData.removeAt(index)
+            statusData.addAll(index, data)
+
+            currentSource?.invalidate()
```

Comment:
it's not obvious to me why it's needed

Reply:
Everytime the data changes, the source needs to be invalidated, which prompts the paging system to create a new source from the source factory. Thats how this library works.

Reply:
We should leave a comment then